File: app/src/main/java/com/example/androidcodingtest/statements/StatementActivity.java
package com.example.androidcodingtest.statements;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.androidcodingtest.login.LoginActivity;
import com.example.androidcodingtest.R;
import com.example.androidcodingtest.models.Statement;
import com.example.androidcodingtest.models.UserAccount;

import java.text.NumberFormat;
import java.util.List;

public class StatementActivity extends AppCompatActivity implements StatementInteractor.View{

    private StatementAdapter mAdapter;
    private UserAccount userAccount;
    private StatementPresenter presenter = new StatementPresenter(this);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_statement);

        ImageView logout = findViewById(R.id.logout);
        logout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getBaseContext(), LoginActivity.class);
                startActivity(intent);
                finish();
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();

        Intent intent = getIntent();
        UserAccount userAccount = getIntent().getParcelableExtra("userAccount");
        if(userAccount != null){
            this.userAccount = userAccount;
            populateUserInfo();
        }
    }

    private void populateUserInfo() {

        TextView userName = findViewById(R.id.user_name);
        TextView userAccountNumber = findViewById(R.id.user_account_number);
        TextView userBalance = findViewById(R.id.user_balance);

        userName.setText(userAccount.getName());
        userAccountNumber.setText(userAccount.getBankAccount() + " / " + userAccount.getAgency());

        NumberFormat format = NumberFormat.getCurrencyInstance();
        userBalance.setText(format.format(userAccount.getBalance()));

        presenter.getStatements(userAccount.getUserId());
    }

    @Override
    public void generateStatementList(List<Statement> statementList) {
        mAdapter = new StatementAdapter(statementList);
        RecyclerView recyclerView = findViewById(R.id.statement_recycler_view);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(mAdapter);
    }

    @Override
    public void showMessage(int message) {
        Toast.makeText(this, getString(message), Toast.LENGTH_SHORT).show();
    }
}

File: BankApp/app/src/main/java/br/com/santander/android/bank/statement/presentation/StatementActivity.kt
package br.com.santander.android.bank.statement.presentation

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import br.com.santander.android.bank.R
import br.com.santander.android.bank.core.extensions.formatAgency
import br.com.santander.android.bank.core.extensions.toCurrency
import br.com.santander.android.bank.login.domain.model.Account
import br.com.santander.android.bank.login.presentation.LoginActivity
import br.com.santander.android.bank.statement.di.StatementInjection
import br.com.santander.android.bank.statement.domain.model.Statement
import br.com.santander.android.bank.statement.domain.model.Statements
import kotlinx.android.synthetic.main.activity_statements.*

internal class StatementActivity : AppCompatActivity(), StatementContract.View {

    private lateinit var account: Account
    private var presenter = StatementPresenter(this, StatementInjection.interactor)
    private var statements: Statements? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_statements)

        if (savedInstanceState != null) {
            account = getAccount(savedInstanceState)
            statements = getStatements(savedInstanceState)
        } else {
            account = getAccount(intent.extras!!)
        }

        setupViews()
        presenter.onCreate()
        presenter.requestStatements(account.userId.toInt())
    }

    private fun setupViews() {
        setSupportActionBar(statements_toolbar)
        val formattedAccount = "${account.bankAccount} / ${account.agency.formatAgency()}"
        txt_name.text = account.name
        txt_account.text = formattedAccount
        txt_balance.text = account.balance.toCurrency()
        btn_exit.setOnClickListener { logout() }
        rv_statements.layoutManager = LinearLayoutManager(this)
        swip_refresh.setOnRefreshListener {
            presenter.requestStatements(account.userId.toInt())
        }
    }

    private fun logout() {
        startActivity(LoginActivity.init(this))
        finish()
    }

    override fun showLoading() {
        swip_refresh.isRefreshing = true
    }

    override fun hideLoading() {
        swip_refresh.isRefreshing = false
    }

    override fun showStatements(statements: Statements) {
        this.statements = statements
        rv_statements.adapter = StatementAdapter(statements.list)
    }

    override fun showTryAgainMessage() {
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putSerializable(ACCOUNT_PARAM_KEY, account)
        outState?.putSerializable(STATEMENTS_PARAM_KEY, statements)
    }

    companion object {
        private const val ACCOUNT_PARAM_KEY = "account_param"
        private const val STATEMENTS_PARAM_KEY = "statements_param"

        fun init(context: Context, account: Account): Intent {
            val intent = Intent(context, StatementActivity::class.java)
            intent.putExtra(ACCOUNT_PARAM_KEY, account)
            return intent
        }

        internal fun getAccount(bundle: Bundle): Account {
            return bundle.getSerializable(ACCOUNT_PARAM_KEY) as Account
        }

        internal fun getStatements(bundle: Bundle): Statements {
            return bundle.getSerializable(STATEMENTS_PARAM_KEY) as Statements
        }
    }
}